migration/migration.go

```diff
@@ -0,0 +1,198 @@
+package migration
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	logger "github.com/ipfs/go-log/v2"
+)
+
+var (
+	log = logger.Logger("migrations")
+
+	keyCurrentVersion = datastore.NewKey("version")
+)
+
+// Migrator ensures a datastore goes through all the needed
+// Migrations to upgrade its current version to the latest version.
+type Migrator struct {
+	ds         datastore.TxnDatastore
+	migrations map[int]Migration
+}
+
+type datastoreReaderWriter interface {
+	datastore.Read
+	datastore.Write
+}
+
+// Migration runs a vA->v(A+1) migration. UseTxn indicates
+// if this migration should be run in a transaction.
+type Migration struct {
+	Run    func(datastoreReaderWriter) error
+	UseTxn bool
+}
```

Comment:
A `Migration` is some logic that does a transformation between version X and X+1.
It also describes if it should run in a txn or not. This is to solve that long/big-running migration can't be transactional in Bager or Mongo (Atlas) for reasons explained in Slack.